### PR TITLE
Backport SM specification to magma for fosscuda/2019b

### PR DESCRIPTION
**Pull request by Flamefire**
(created using `eb --new-pr`)

### CODE REVIEW

**Comment by Flamefire** [comment]
Test report by @Flamefire
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
taurusml19 - Linux RHEL 7.6, POWER, 8335-GTX, Python 2.7.5
See https://gist.github.com/e0872ac0c6ad07f3fc9772b08aeffa3d for a full test report.

**Comment by Flamefire** [comment]
Test report by @Flamefire
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
taurusa6 - Linux centos linux 7.7.1908, x86_64, Intel(R) Xeon(R) CPU E5-2603 v4 @ 1.70GHz, Python 2.7.5
See https://gist.github.com/81230cbdd019127ee05398f9ad424dd6 for a full test report.

**Comment by Flamefire** [comment]
FTR: I submitted the patch I added upstream as https://bitbucket.org/icl/magma/pull-requests/5. If you want we can hold it back and wait for the "official" patch or approval of mine.

**Comment by lexming** [comment]
Test report by @lexming
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
node156.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz, Python 2.7.5
See https://gist.github.com/6bceaafa0193dfaff81f7694da16b081 for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
node156.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz, Python 2.7.5
See https://gist.github.com/331571a394822760742b5aaf90da8221 for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
node154.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz, Python 2.7.5
See https://gist.github.com/d099c931236eaa175e980e8c1515efef for a full test report.

File: easybuild/easyconfigs/m/magma/magma-2.5.4-fosscuda-2019b.eb
Comment (lexming) [inline easybuild/easyconfigs/m/magma/magma-2.5.4-fosscuda-2019b.eb:25]: This is not working without `cuda_compute_capabilities` being set. Instead of falling back to `Kepler Maxwell Pascal Volta`, the configuration has `-DGPU_TARGET="%(cuda_sm_space_sep)s"`.
I guess that in this case the template `cuda_sm_space_sep` has to be set to an empty string instead of being undefined.

**Review by lexming — CHANGES_REQUESTED** [review]

File: easybuild/easyconfigs/m/magma/magma-2.5.4-fosscuda-2019b.eb
Comment (lexming) [reply easybuild/easyconfigs/m/magma/magma-2.5.4-fosscuda-2019b.eb:25]: the issue might be that the templates are replaced later on, once the easyconfig is fully parsed. So this will never work as intended regardless of the contents of `%(cuda_sm_space_sep)s`.

Comment (Flamefire) [reply easybuild/easyconfigs/m/magma/magma-2.5.4-fosscuda-2019b.eb:25]: Hm, any idea how to generically solve this? This should likely go into the framework: Use build_option OR EC option for `cuda_compute_capabilities`. Currently this is implemented (twice!) in PyTorch and Tensorflow EasyBlocks and now we need it a third time.
ccing @boegel

Comment (Flamefire) [reply easybuild/easyconfigs/m/magma/magma-2.5.4-fosscuda-2019b.eb:25]: Oh, dismiss my last comment, I already implemented it like that. See the update to this PR, this should work now.
I only had to use the magma sm assigned values for the names

File: easybuild/easyconfigs/m/magma/magma-2.5.4-fosscuda-2019b.eb
Comment (Micket) [inline easybuild/easyconfigs/m/magma/magma-2.5.4-fosscuda-2019b.eb:26]: I think it would be best if this list of version numbers matched the string, so, 7.0 for Volta should be included as well. 
Also, while we are at it, could we also add `'Turing'` and `7.5` in the defaults?
```suggestion
cuda_compute_capabilities = ['3.0', '3.5', '5.0', '6.0', '7.0', '7.5']  # or 'Kepler Maxwell Pascal Volta Turing'
```

**Review by Micket — CHANGES_REQUESTED** [review]

File: easybuild/easyconfigs/m/magma/magma-2.5.4-fosscuda-2019b.eb
Comment (Flamefire) [reply easybuild/easyconfigs/m/magma/magma-2.5.4-fosscuda-2019b.eb:26]: Good catch. The magma default only adds Pascal and that is where I had the default from.
I wouldn't add Turing though as this is an existing EC so I'd keep the changes minimal. But no strong opinion from my side.

Comment (Micket) [reply easybuild/easyconfigs/m/magma/magma-2.5.4-fosscuda-2019b.eb:26]: There is no harm in adding this I think? Noone is required to rebuild, but it's just better default for those who are "late to the party". Besides, it would already change for anyone who uses `--cuda-compute-capabilities`. Since the test-build is quite long, I would rather add it directly in this PR.

Another thought i had was; The "string" version would not be useable here at all right? I can only imagine it totally breaking `cuda_compute_capabilities`. Do we even need it?

Comment (Flamefire) [reply easybuild/easyconfigs/m/magma/magma-2.5.4-fosscuda-2019b.eb]: Done.
And yes that would not be usable. I only left it there as a comment so people know what that is/was. I'd remove the comment in new magma ECs

**Review by Micket — APPROVED** [review]
lgtm

**Review by Micket — APPROVED** [review]
lgtm

**Review by lexming — APPROVED** [review]
LGTM

**Comment by lexming** [comment]
Test report by @lexming
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
node154.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz, Python 2.7.5
See https://gist.github.com/f162cfb1e3c6b4fdc299c08901f0b822 for a full test report.

**Comment by Micket** [comment]
I'm guessing disk space @lexming (at least that's the usual culprit). (I'm still building toolchains, magma build coming up soon)

**Comment by Flamefire** [comment]
Maybe we should start to check for free disk space, especially for software like magma or TensorFlow which consume 2-3 digit GBs of space

**Comment by Micket** [comment]
I just started a build, and I'm ending up with;
```
--     NVCCFLAGS:             -Xcompiler -fPIC;-gencode arch=compute_61,code=sm_61;-gencode arch=compute_70,code=sm_70;-gencode arch=compute_75,code=sm_75;-gencode arch=compute_75,code=compute_75
```
that can't be right?

**Comment by lexming** [comment]
It's not free disk space
```
nvcc fatal   : Unknown option 'Xcompiler -fPIC'
```
I think that is caused by the patch

File: easybuild/easyconfigs/m/magma/magma-2.5.4_allow-all-sms.patch
Comment (lexming) [inline easybuild/easyconfigs/m/magma/magma-2.5.4_allow-all-sms.patch:181]: This is causing the failed tests

**Review by lexming — DISMISSED** [review]

**Comment by Flamefire** [comment]
@Micket That should be correct. According to the docs the values should be semicolon separated.

I'm curious how:
`NVCCFLAGS:             -Xcompiler -fPIC;-gencode arch=compute_61,code=sm_61;-gencode arch=compute_70,code=sm_70`
ends up as 
`nvcc fatal   : Unknown option 'Xcompiler -fPIC'`

It works for compiling but linking has `nvcc "-Xcompiler -fPIC" "-gencode arch=compute_30,code=sm_30"`...

So that looks like there should be more semicolons. I'll look at that tomorrow

**Comment by lexming** [comment]
@Flamefire indeed, the problem seems to be the quotes, they are all preserved in the command

> /bin/nvcc "-Xcompiler -fPIC" "-gencode arch=compute_30,code=sm_30" "-gencode arch=compute_35,code=sm_35" "-gencode arch=compute_50,code=sm_50" "-gencode arch=compute_60,code=sm_60" "-gencode arch=compute_70,code=sm_70" "-gencode arch=compute_75,code=sm_75" "-gencode arch=compute_75,code=compute_75" -m64 -ccbin /apps/brussel/CO7/ivybridge-ib/software/GCCcore/8.3.0/bin/gcc -dlink [...]

by removing all quotes it works.

**Comment by Flamefire** [comment]
> by removing all quotes it works.

While you are at it: Do the quotes around the gencode stuff work or do they also have to be removed?

**Comment by lexming** [comment]
@Flamefire quotes around `gencode` have to be removed as well

File: easybuild/easyconfigs/m/magma/magma-2.5.4_allow-all-sms.patch
Comment (Flamefire) [reply easybuild/easyconfigs/m/magma/magma-2.5.4_allow-all-sms.patch]: Fixed

**Comment by Flamefire** [comment]
Test report by @Flamefire
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
taurusml22 - Linux RHEL 7.6, POWER, 8335-GTX, Python 2.7.5
See https://gist.github.com/ff0f26fb15cdaa4635eef3b45ed5fc48 for a full test report.

**Comment by Flamefire** [comment]
Test report by @Flamefire
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
taurusa15 - Linux centos linux 7.7.1908, x86_64, Intel(R) Xeon(R) CPU E5-2603 v4 @ 1.70GHz, Python 2.7.5
See https://gist.github.com/8174d101f0742ea0f5c470551439bf82 for a full test report.

**Comment by Micket** [comment]
Test report by @Micket
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
vera-c1 - Linux centos linux 7.9.2009, x86_64, Intel Xeon Processor (Skylake), Python 3.6.8
See https://gist.github.com/4236d2d658d50f0b610903748a5dfd0b for a full test report.

File: easybuild/easyconfigs/m/magma/magma-2.5.4-fosscuda-2019b.eb
Comment (Micket) [inline easybuild/easyconfigs/m/magma/magma-2.5.4-fosscuda-2019b.eb:26]: I don't you intended to remove 7.0 and 7.5 again?

**Review by Micket — CHANGES_REQUESTED** [review]

File: easybuild/easyconfigs/m/magma/magma-2.5.4-fosscuda-2019b.eb
Comment (Flamefire) [reply easybuild/easyconfigs/m/magma/magma-2.5.4-fosscuda-2019b.eb:26]: Woops. update-pr is tricky, thanks!

**Review by Micket — APPROVED** [review]
lgtm

**Comment by Micket** [comment]
Test report by @Micket
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
vera-c1 - Linux centos linux 7.9.2009, x86_64, Intel Xeon Processor (Skylake), Python 3.6.8
See https://gist.github.com/27db43a9fe6e6757cfc465f06e715c6b for a full test report.

**Comment by Micket** [comment]
Going in, thanks @Flamefire!

**Comment by Micket** [comment]
Going in, thanks @Flamefire!

**Comment by lexming** [comment]
Test report by @lexming
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
node156.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz, Python 2.7.5
See https://gist.github.com/6a627dc2918d66a4f21b084a5683f539 for a full test report.

**Pull request merged**
